src/rt/monitor_.d

```diff
@@ -1,444 +1,289 @@
-        assert(m.impl is null);
-
-        foreach (ref v; m.devt)
-        {
-            if (v is null || v == e)
-            {
-                v = e;
-                return;
-            }
-        }
-
-        auto len = m.devt.length + 4; // grow by 4 elements
-        auto pos = m.devt.length;     // insert position
-        auto p = realloc(m.devt.ptr, DEvent.sizeof * len);
-        import core.exception : onOutOfMemoryError;
-        if (!p)
-            onOutOfMemoryError();
-        m.devt = (cast(DEvent*)p)[0 .. len];
-        m.devt[pos+1 .. len] = null;
-        m.devt[pos] = e;
-    }
-}
-
-extern (C) void rt_detachDisposeEvent(Object h, DEvent e)
-{
-    synchronized (h)
-    {
-        Monitor* m = getMonitor(h);
-        assert(m.impl is null);
-
-        foreach (p, v; m.devt)
-        {
-            if (v == e)
-            {
-                memmove(&m.devt[p],
-                        &m.devt[p+1],
-                        (m.devt.length - p - 1) * DEvent.sizeof);
-                m.devt[$ - 1] = null;
-                return;
-            }
-        }
-    }
-}
+import core.atomic, core.stdc.stdlib, core.stdc.string;
 
-nothrow:
+// NOTE: The dtor callback feature is only supported for monitors that are not
+//       supplied by the user.  The assumption is that any object with a user-
+//       supplied monitor may have special storage or lifetime requirements and
+//       that as a result, storing references to local objects within Monitor
+//       may not be safe or desirable.  Thus, devt is only valid if impl is
+//       null.
 
-private
+extern (C) void _d_setSameMutex(shared Object ownee, shared Object owner) nothrow in
 {
-    debug(PRINTF) import core.stdc.stdio;
-    import core.stdc.stdlib;
-    import core.stdc.string;
-    import core.atomic;
-
-    version( CRuntime_Glibc )
-    {
-        version = USE_PTHREADS;
-    }
-    else version( FreeBSD )
-    {
-        version = USE_PTHREADS;
-    }
-    else version( OSX )
+    assert(ownee.__monitor is null);
+}
+body
+{
+    auto m = ensureMonitor(cast(Object) owner);
+    auto i = m.impl;
+    if (i is null)
     {
-        version = USE_PTHREADS;
+        atomicOp!("+=")(m.refs, cast(size_t) 1);
+        ownee.__monitor = owner.__monitor;
+        return;
     }
-    else version( Solaris )
+    // If m.impl is set (ie. if this is a user-created monitor), assume
+    // the monitor is garbage collected and simply copy the reference.
+    ownee.__monitor = owner.__monitor;
+}
+
+extern (C) void _d_monitordelete(Object h, bool det)
+{
+    auto m = getMonitor(h);
+    if (m is null)
+        return;
+
+    if (m.impl)
     {
-        version = USE_PTHREADS;
+        // let the GC collect the monitor
+        setMonitor(h, null);
     }
-    else version( CRuntime_Bionic )
+    else if (!atomicOp!("-=")(m.refs, cast(size_t) 1))
     {
-        version = USE_PTHREADS;
+        // refcount == 0 means unshared => no synchronization required
```

Comment:
It's safe because you can't call setSameMutex on an object whose refcount just dropped to 0, or otherwise it wouldn't be collected, meaning you no longer own it.